src/EventStore.Core/Services/Archive/Archiver/ArchiverService.cs

```diff
@@ -3,240 +3,116 @@
 
 using System;
 using System.Collections.Concurrent;
-using System.Collections.Generic;
-using System.Diagnostics;
-using System.IO;
-using System.Linq;
+using System.Runtime.CompilerServices;
 using System.Threading;
-using System.Threading.Channels;
 using System.Threading.Tasks;
+using DotNext.Runtime.CompilerServices;
+using DotNext.Threading;
 using EventStore.Core.Bus;
 using EventStore.Core.Data;
 using EventStore.Core.Messages;
 using EventStore.Core.Services.Archive.Storage;
-using EventStore.Core.Services.Archive.Storage.Exceptions;
-using EventStore.Core.Services.Archive.Archiver.Unmerger;
-using Serilog;
+using EventStore.Core.TransactionLog.Chunks;
+using EventStore.Core.TransactionLog.Chunks.TFChunk;
 
 namespace EventStore.Core.Services.Archive.Archiver;
 
-public class ArchiverService :
-	IHandle<SystemMessage.ChunkLoaded>,
-	IHandle<SystemMessage.ChunkCompleted>,
+public sealed class ArchiverService :
+	IHandle<SystemMessage.SystemStart>,
+	IHandle<SystemMessage.BecomeShuttingDown>,
 	IHandle<SystemMessage.ChunkSwitched>,
 	IHandle<ReplicationTrackingMessage.ReplicatedTo>,
-	IHandle<SystemMessage.BecomeShuttingDown>
-{
-	private static readonly ILogger Log = Serilog.Log.ForContext<ArchiverService>();
+	IAsyncDisposable {
 
-	private readonly ISubscriber _mainBus;
 	private readonly IArchiveStorage _archive;
-	private readonly Queue<ChunkInfo> _uncommittedChunks;
-	private readonly ConcurrentDictionary<string, ChunkInfo> _existingChunks;
-	private readonly CancellationTokenSource _cts;
-	private readonly Channel<Commands.ArchiveChunk> _archiveChunkCommands;
-	private readonly IChunkUnmerger _chunkUnmerger;
+	private readonly CancellationToken _lifetimeToken;
+	private readonly AsyncAutoResetEvent _archivingSignal;
+	private readonly IChunkRegistry<IChunkBlob> _chunkManager;
+	private readonly ConcurrentBag<ChunkInfo> _switchedChunks;
```

Comment:
would it be better if we use a set data structure, using the `ChunkEndNumber` as key?

1. it would ensure we can iterate on the switched chunks in order when archiving them
2. consider the situation where the archive storage is down for a long time (e.g a few days). in that case `_archive.StoreChunk` will take a long time to complete. in the mean time, it may have been possible that the archive was scavenged several times. so we would be uploading the scavenged chunks multiple times when the archive storage comes back up. with the set data structure, there wouldn't be duplicates.

Reply:
> it would ensure we can iterate on the switched chunks in order when archiving them

The order still not guaranteed. The background job can take switched chunk `1` to upload it, in the same time `ChunkSwitched` can happen concurrently and add chunk `0`.

Reply:
it's a good point about not uploading multiple times though

Reply:
actually this becomes a bit complicated, lets leave the deduplication out for now and reconsider it when we figure out the details of scavenging the archive